services/requester/batch_tx_pool.go

```diff
@@ -0,0 +1,286 @@
+
+	// tx sending should be blocking, so we don't have races when
+	// pooled transactions are being processed in the background.
+	t.txMux.Lock()
+	defer t.txMux.Unlock()
+
+	from, err := gethTypes.Sender(gethTypes.LatestSignerForChainID(tx.ChainId()), tx)
+	if err != nil {
+		return fmt.Errorf("failed to derive the sender: %w", err)
+	}
+	txData, err := tx.MarshalBinary()
+	if err != nil {
+		return err
+	}
+	hexEncodedTx, err := cadence.NewString(hex.EncodeToString(txData))
+	if err != nil {
+		return err
+	}
+
+	userTx := pooledEvmTx{txPayload: hexEncodedTx, nonce: tx.Nonce()}
+	t.pooledTxs[from] = append(t.pooledTxs[from], userTx)
+
+	return nil
+}
+
+func (t *BatchTxPool) processPooledTransactions() {
+	ticker := time.NewTicker(t.config.TxBatchInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-t.ctx.Done():
+			return
+		case <-ticker.C:
+			latestBlock, account, err := t.fetchFlowLatestBlockAndCOA()
+			if err != nil {
+				t.logger.Error().Err(err).Msg(
+					"failed to get COA / latest Flow block on batch tx submission",
+				)
+				continue
+			}
+
+			snapshot := func() map[gethCommon.Address][]pooledEvmTx {
+				t.txMux.Lock()
+				defer t.txMux.Unlock()
+				poolCopy := make(map[gethCommon.Address][]pooledEvmTx)
+				for addr, txs := range t.pooledTxs {
+					poolCopy[addr] = make([]pooledEvmTx, len(txs))
+					copy(poolCopy[addr], txs)
+				}
+				t.pooledTxs = make(map[gethCommon.Address][]pooledEvmTx)
+				return poolCopy
+			}
+			// Take a snapshot here to allow `Add()` to continue accept
+			// incoming EVM transactions, without blocking until the
+			// batch transactions are submitted.
+			txsGroupedByAddress := snapshot()
+
+			for address, pooledTxs := range txsGroupedByAddress {
+				err := t.batchSubmitTransactionsForSameAddress(
+					latestBlock,
+					account,
+					pooledTxs,
+				)
+				if err != nil {
+					t.logger.Error().Err(err).Msgf(
+						"failed to send Flow transaction from BatchPool for EOA: %s",
+						address.Hex(),
+					)
+					continue
+				}
+			}
+		}
+	}
+}
+
+func (t *BatchTxPool) batchSubmitTransactionsForSameAddress(
+	latestBlock *flow.Block,
+	account *flow.Account,
+	pooledTxs []pooledEvmTx,
+) error {
+	// Sort the transactions based on their nonce, to make sure
+	// that no re-ordering has happened due to races etc.
+	sort.Slice(pooledTxs, func(i, j int) bool {
+		return pooledTxs[i].nonce < pooledTxs[j].nonce
+	})
+
+	hexEncodedTxs := make([]cadence.Value, len(pooledTxs))
+	for i, txPayload := range pooledTxs {
+		hexEncodedTxs[i] = txPayload.txPayload
+	}
+
+	coinbaseAddress, err := cadence.NewString(t.config.Coinbase.Hex())
+	if err != nil {
+		return err
+	}
+
+	script := replaceAddresses(batchRunTxScript, t.config.FlowNetworkID)
+	flowTx, err := t.buildTransaction(
+		latestBlock,
+		account,
+		script,
+		cadence.NewArray(hexEncodedTxs),
+		coinbaseAddress,
+	)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
```

Comment:
use the context passed to `processPooledTransactions()` instead of `context.Background()`?

Reply:
Good point :+1: Updated in https://github.com/onflow/flow-evm-gateway/pull/831/commits/2a98db83c0099a71511819ff0d75d3747fde5700 .
Also removed the `WithTimeout` addition, since `SingleTxPool` didn't use anything like that, and the timeout I added was a bit magic.